services/galley/src/Galley/API/Public.hs

```diff
@@ -72,6 +95,239 @@ import qualified Wire.API.Team.SearchVisibility as Public
 import qualified Wire.API.User as Public (UserIdList, modelUserIdList)
 import Wire.Swagger (int32Between)
 
+-- This type exists for the special 'HasSwagger' and 'HasServer' instances. It
+-- shows the "Authorization" header in the swagger docs, but expects the
+-- "Z-Auth" header in the server. This helps keep the swagger docs usable
+-- through nginz.
+data ZUserType = ZAuthUser | ZAuthConn
+
+type family ZUserHeader (ztype :: ZUserType) :: Symbol where
+  ZUserHeader 'ZAuthUser = "Z-User"
+  ZUserHeader 'ZAuthConn = "Z-Connection"
+
+type family ZUserParam (ztype :: ZUserType) :: * where
+  ZUserParam 'ZAuthUser = UserId
+  ZUserParam 'ZAuthConn = ConnId
+
+data ZAuthServant (ztype :: ZUserType)
+
+type InternalAuth ztype =
+  Header'
+    '[Servant.Required, Servant.Strict]
+    (ZUserHeader ztype)
+    (ZUserParam ztype)
+
+type ZUser = ZAuthServant 'ZAuthUser
+
+instance HasSwagger api => HasSwagger (ZAuthServant 'ZAuthUser :> api) where
+  toSwagger _ =
+    toSwagger (Proxy @api)
+      & securityDefinitions <>~ InsOrdHashMap.singleton "ZAuth" secScheme
+      & security <>~ [SecurityRequirement $ InsOrdHashMap.singleton "ZAuth" []]
+    where
+      secScheme =
+        SecurityScheme
+          { _securitySchemeType = SecuritySchemeApiKey (ApiKeyParams "Authorization" ApiKeyHeader),
+            _securitySchemeDescription = Just "Must be a token retrieved by calling 'POST /login' or 'POST /access'. It must be presented in this format: 'Bearer \\<token\\>'."
+          }
+
+instance HasSwagger api => HasSwagger (ZAuthServant 'ZAuthConn :> api) where
+  toSwagger _ = toSwagger (Proxy @api)
+
+instance
+  ( HasContextEntry (ctx .++ DefaultErrorFormatters) ErrorFormatters,
+    HasServer api ctx,
+    KnownSymbol (ZUserHeader ztype),
+    FromHttpApiData (ZUserParam ztype)
+  ) =>
+  HasServer (ZAuthServant ztype :> api) ctx
+  where
+  type ServerT (ZAuthServant ztype :> api) m = ServerT (InternalAuth ztype :> api) m
+
+  route _ = Servant.route (Proxy @(InternalAuth ztype :> api))
+  hoistServerWithContext _ pc nt s =
+    Servant.hoistServerWithContext (Proxy @(InternalAuth ztype :> api)) pc nt s
+
+instance ToSchema a => ToSchema (Headers ls a) where
+  declareNamedSchema _ = declareNamedSchema (Proxy @a)
+
+data Api routes = Api
+  { -- Conversations
+
+    getConversation ::
+      routes
+        :- Summary "Get a conversation by ID"
+        :> ZUser
+        :> "conversations"
+        :> Capture "cnv" ConvId
+        :> Get '[Servant.JSON] Public.Conversation,
+    getConversationRoles ::
+      routes
+        :- Summary "Get existing roles available for the given conversation"
+        :> ZUser
+        :> "conversations"
+        :> Capture "cnv" ConvId
+        :> Get '[Servant.JSON] Public.ConversationRolesList,
+    getConversationIds ::
+      routes
+        :- Summary "Get all conversation IDs."
+        -- TODO: add note "At most 1000 IDs are returned per request"
```

Comment:
Does this not get added automatically due to the `ToSchema` instance of `Range`?

Reply:
Not yet, but good point, that instance could include the bounds. I'll add a note.